docs/adapter_impl_notes.md

```diff
@@ -0,0 +1,184 @@
+
+## Fields and Searching
+
+Every ``read`` command can contain a [filter expression](https://github.com/juttle/juttle/blob/master/docs/concepts/filtering.md) that is used as an initial filter for the messages selected by the adapter. An adapter is not obligated to support a filter expression. The ``filter`` proc allows for filtering of points within programs. However, for performance reasons it is highly recommended that adapters implement filter expressions and push filtering into the backend whenever possible.
+
+A filter expression either takes the form of a free-text search or a field match expression, possibly combined with logical operators like ``AND``, ``OR``, etc.
+
+## The ``read`` Proc
+
+The ``read`` proc is the interface between the juttle program and the adapter implementation. The specific format of the options to ``read`` are not enforced by the Juttle compiler. A read proc can have any number of options. However, there are conventions for commonly supported options:
+
+```
+read <adapter> [-from <moment>] [-to <moment>] [-timeField <field>] [-raw <expression>] [<filter expression>]
+```
+
+* ``-from``/``-to``: [Juttle Moments](https://github.com/juttle/juttle/blob/master/lib/moment/juttle-moment.js) representing the start and end of the time range for the read.
+* ``-timeField``: A field from the backend data that should be used as the time of the points emitted to the program.
+* ``-raw``:
+
+Let's discuss how the Gmail adapter interprets ``-from``, ``-to``, ``-raw``, and the filter expression to select messages:
+
+The Gmail api supports date-based searches via ``before:`` and ``after:``. However, the arguments to ``before:`` and ``after:`` can only be dates, while the ``-from``/``-to`` options to ``read`` have greater (sub-second) precision. So when fetching messages, the adapter rounds down the ``-from`` to the beginning of the day and ``-to`` to the end of the day. Afterward, the adapter compares the actual message receipt time (in the ``internalDate`` field) against the ``-from``/``-to`` and only passes matching messages along to ``emit()``.
+
+Field matches in search expressions are interpreted as message header matches for a limited set of headers:
+
+* ``from``
+* ``to``
+* ``subject``
+* ``cc``
+* ``bcc``
+
+The following comparison operators are supported:
+
+* ~, =~ (wildcard operator). This is because Gmail's heaader searches match on substrings and do not perform exact matches.
+* !~ (wildcard negation).
+
+These header matches are pushed into the Gmail API search expression. Logical operators such as ``AND``, ``OR``, and ``NOT`` join terms in the expression.
+
+Free-text search is supported by the Gmail API, so any free-text searches are passed through to the search expression.
```

Comment:
Is full-text search done on the message body, or across all headers and message body? Looks like Gmail API lets you do search on a given field, such as `subject:"dinner and a movie"` (and defaults to searching all? wasn't clear to me from their doc), but Juttle currently does not, our FTS is supposed to match values of any field.